### PR TITLE
KTime: Sync with realtime clock

### PR DESCRIPTION
**Pull request by christos68k**
### Summary

- Moved `KTime` / `GetKTime` from `util` to `times` package
- Periodic sync with realtime clock (instead of boottime monotonic)
- Added realtime clock sync interval configuration option

### CODE REVIEW

File: times/times.go [hunk 1]
```diff
@@ -8,30 +8,19 @@ package times
...
-	// In the kernel, we retrieve timestamps from a monotonic clock
-	// (bpf_ktime_get_ns) that does not count system suspend time.
-	// In userspace, we try to detect system suspend events by diffing
-	// with values retrieved from a monotonic clock that does count
-	// the system suspend time.
-	// If the delta exceeds the following threshold (10s), we add it to the
-	// monotonic 'delta' that we keep track of.
-	monotonicThresholdNs = 10 * 1e9
+	// How often to attempt a sync with the realtime clock.
+	realtimeSyncInterval = 20 * time.Minute
```
Comment (christos68k) [inline times/times.go:23]: There are multiple options for realtime clock sync:

1. Sync once, on agent startup (calculate the delta once which remains fixed for the life of the agent process).
2. Sync periodically but do not allow time to drift backwards (only update delta if `newDelta - oldDelta > 0`)
3. Sync periodically (this PR).

Option 3 is more accurate in reflecting realtime clock changes but can result in discontinuous (non-monotonic) timestamps.

**Comment by rockdaboot** [comment]
Gave it a spin with the latest changes.
When after putting my machine to sleep for ~10min... the timestamps after wake-up are off by these 10mins. I guess until the sync kicks in... Maybe we can make the sync every 1min or is this eating too much CPU in comparison to the total CPU usage of the agent (which is very low)?

**Comment by christos68k** [comment]
> I guess until the sync kicks in... Maybe we can make the sync every 1min or is this eating too much CPU in comparison to the total CPU usage of the agent (which is very low)?

It's a tradeoff, but not really for CPU usage: higher sync frequency increases the time window where timestamps can be off relative to realtime clock but lower sync frequency also increases possible number of discontinuities (e.g. timestamps jumping around). We should probably make sync frequency configurable and allow for it to be zero (in which case we're reverting to [Option 1](https://github.com/elastic/otel-profiling-agent/pull/96#discussion_r1700275818)).

**Comment by rockdaboot** [comment]
> > I guess until the sync kicks in... Maybe we can make the sync every 1min or is this eating too much CPU in comparison to the total CPU usage of the agent (which is very low)?
> 
> It's a tradeoff, but not really for CPU usage: higher sync frequency increases the time window where timestamps can be off relative to realtime clock but lower sync frequency also increases possible number of discontinuities (e.g. timestamps jumping around). We should probably make sync frequency configurable and allow for it to be zero (in which case we're reverting to [Option 1](https://github.com/elastic/otel-profiling-agent/pull/96#discussion_r1700275818)).

Having a configurable value sounds good to me.
Is it a common issue that timestamps are jumping back and forth as you say? I can't remember ever having an issue with that. Jumps should be relatively rare under normal conditions.

Meanwhile, the sync kicked in. I have ~10mins of wrong time stamps and a gap in my time series at a wrong and unexpected time range.

So I wonder if we can't detect jumps in time faster and only resync when required. Wasn't that what the current code is doing?

**Comment by christos68k** [comment]
> So I wonder if we can't detect jumps in time faster and only resync when required. Wasn't that what the current code is doing?

The current code is more complicated and the consensus in previous discussions was to drastically simplify. We can certainly add threshold logic ([Option 3](https://github.com/elastic/otel-profiling-agent/pull/96#discussion_r1700275818)) if the current consensus is that it's worth it. 

My preference is to make sync-frequency configurable and lower the frequency e.g. (2-3 minutes).

IIRC a few other Go agents that deal with monotonic timestamps that I've looked at only sync once and keep the delta fixed for the life of the agent.

**Comment by rockdaboot** [comment]
> My preference is to make sync-frequency configurable and maybe reduce the default value (e.g. 2-3 minutes).

Sounds pragmatic, 2-3 minutes is acceptable at least to me.

Maybe add a short section to the README to explain the timestamp issue after sleep/power-saving!?

File: times/times.go [hunk 2]
```diff
@@ -123,15 +107,22 @@ func (t *Times) PIDCleanupInterval() time.Duration { return t.pidCleanupInterval
...
+	})
 }
 
 // New returns a new Times instance.
 func New(ctx context.Context, reportInterval, monitorInterval,
 	probabilisticInterval time.Duration) *Times {
-	// Start periodic synchronization of the monotonic clock
-	StartMonotonicSync(ctx)
+	// Start periodic synchronization with the realtime clock
+	StartRealtimeSync(ctx)
```
[the remark this answers is not in the document]
Comment (christos68k) [reply times/times.go]: I decoupled it from `times.New` and it's now called from `main`.

File: times/times.go (hunk 1 above)
Comment (rockdaboot) [reply times/times.go:23]: Can you amend this to 2 or 3 minutes and make it configurable, following your suggestion? For me it's fine if we do the latter in a follow-up PR.

**Review by rockdaboot — APPROVED** [review]
LGTM

File: times/ktime.go [hunk 3]
```diff
@@ -4,9 +4,10 @@
  * See the file "LICENSE" for details.
  */
 
-package util
+package times
```
[the remark this answers is not in the document]
Comment (athre0z) [reply times/ktime.go:7]: Hmm, yeah. I agree that this is a weird mixture in this package, but I don't see it as  this PR's fault: the monotonic sync code was in this package already before. Previously some functions just happened to **additionally** be scattered in `util`. We should probably split this package up in a separate PR.

**Review by athre0z — APPROVED** [review]
LGTM!

**Pull request merged**